### PR TITLE
VNet/Vxlan delete handling

### PR DESCRIPTION
**Pull request by prsunny**
**What I did**
Delete handling for VNet Routes, NH tunnels and VNets

**Why I did it**
To support deletion scenarios

**How I verified it**
Tested on DUT. VS test case would be added subsequently

**Details if related**
Based on design https://github.com/Azure/SONiC/pull/324

### CODE REVIEW

File: orchagent/intfsorch.cpp [hunk 1]
```diff
@@ -155,6 +155,7 @@ bool IntfsOrch::setIntf(const string& alias, sai_object_id_t vrf_id, const IpPre
         return false;
     }
 
+    vrf_id = port.m_vr_id;
```
Comment (marian-pritsak) [inline orchagent/intfsorch.cpp]: Can you explain why is this required?

File: orchagent/vnetorch.cpp [hunk 2]
```diff
@@ -15,21 +15,23 @@
...
+extern NeighOrch *gNeighOrch;
 
 /*
  * VRF Modeling and VNetVrf class definitions
  */
 std::vector<VR_TYPE> vr_cntxt;
 
-VNetVrfObject::VNetVrfObject(const std::string& vnet, string& tunnel, set<string>& peer,
-                             vector<sai_attribute_t>& attrs) : VNetObject(tunnel, peer)
+VNetVrfObject::VNetVrfObject(const std::string& vnet, string& tunnel, set<string>& peer, uint32_t vni,
```
Comment (marian-pritsak) [inline orchagent/vnetorch.cpp]: Maybe it is worth considering grouping all these parameters into a single struct VnetInfo?

Comment (prsunny) [reply orchagent/vnetorch.cpp]: Addressed

File: orchagent/vnetorch.cpp [hunk 3]
```diff
@@ -302,16 +411,64 @@ bool VNetOrch::delOperation(const Request& request)
...
+
+            if (vrf_obj->getRouteCount())
+            {
+                SWSS_LOG_ERROR("VNET '%s': Routes are still present", vnet_name.c_str());
+                return false;
+            }
+
+            if (!vxlan_orch->removeVxlanTunnelMap(vrf_obj->getTunnelName(), vrf_obj->getVni()))
+            {
+                SWSS_LOG_ERROR("VNET '%s' map delete failed", vnet_name.c_str());
```
Comment (lguohan) [inline orchagent/vnetorch.cpp]: >SWSS_LOG_ERROR("VNET '%s' map delete failed", vnet_name.c_str()); [](start = 16, length = 65)

why no return false here?

File: orchagent/vnetorch.cpp [hunk 4]
```diff
@@ -426,17 +575,43 @@ bool VNetRouteOrch::doRouteTask<VNetVrfObject>(const string& vnet, IpPrefix& ipP
...
-    if (!gPortsOrch->getPort(ifname, p) || (p.m_rif_id == SAI_NULL_OBJECT_ID))
+    auto *vrf_obj = vnet_orch_->getTypePtr<VNetVrfObject>(vnet);
+    if (op == DEL_COMMAND && !vrf_obj->getRouteNH(ipPrefix, nh))
     {
-        SWSS_LOG_WARN("Port/RIF %s doesn't exist", ifname.c_str());
+        SWSS_LOG_WARN("VNET %s, Route %s get NH failed", vnet.c_str(), ipPrefix.to_string().c_str());
+        return true;
+    }
+
+    bool subnet = (!nh.ips.getSize())?true:false;
```
Comment (lguohan) [inline orchagent/vnetorch.cpp]: >subnet [](start = 9, length = 6)

->isSubnet

File: orchagent/vnetorch.cpp [hunk 5]
```diff
@@ -481,23 +697,31 @@ void VNetRouteOrch::handleRoutes(const Request& request)
...
 
     if (vnet_orch_->isVnetExecVrf())
     {
-        if (!doRouteTask<VNetVrfObject>(vnet_name, ip_pfx, ifname))
+        if (!doRouteTask<VNetVrfObject>(vnet_name, ip_pfx, nh, op))
         {
-            throw std::runtime_error("Route add failed");
+            throw std::runtime_error("Route update failed");
         }
     }
```
Comment (lguohan) [inline orchagent/vnetorch.cpp]: what is else condition? do we need to print warning, log, ???

File: orchagent/vnetorch.cpp [hunk 6]
```diff
@@ -481,23 +697,31 @@ void VNetRouteOrch::handleRoutes(const Request& request)
...
+    SWSS_LOG_INFO("VNET-RT '%s' op '%s' for ip %s", vnet_name.c_str(),
+                   op.c_str(), ip_pfx.to_string().c_str());
 
     if (vnet_orch_->isVnetExecVrf())
     {
-        if (!doRouteTask<VNetVrfObject>(vnet_name, ip_pfx, ifname))
+        if (!doRouteTask<VNetVrfObject>(vnet_name, ip_pfx, nh, op))
         {
-            throw std::runtime_error("Route add failed");
+            throw std::runtime_error("Route update failed");
```
Comment (lguohan) [inline orchagent/vnetorch.cpp]: >  throw std::runtime_error("Route update failed"); [](start = 10, length = 50)

it looks to me it is dangerous to crash here. what if the nexthop interface is not created yet in the doRouteTask.

Comment (lguohan) [reply orchagent/vnetorch.cpp]: I checked the routeorch.cpp, it does not throw error, if add or delete route fails.

---
In reply to: [253250419](https://github.com/Azure/sonic-swss/pull/766#discussion_r253250419) [](ancestors = 253250419)

Comment (lguohan) [reply orchagent/vnetorch.cpp]: it just retry.

---
In reply to: [253250430](https://github.com/Azure/sonic-swss/pull/766#discussion_r253250430) [](ancestors = 253250430,253250419)

File: orchagent/vnetorch.cpp [hunk 7]
```diff
@@ -122,6 +124,112 @@ bool VNetVrfObject::updateObj(vector<sai_attribute_t>& attrs)
...
+    }
+    return true;
+}
+
+size_t VNetVrfObject::getRouteCount() const
+{
+    return (routes_.size() + tunnels_.size());
+}
+
+bool VNetVrfObject::getRouteNH(IpPrefix& ipPrefix, nextHop& nh)
```
Comment (lguohan) [inline orchagent/vnetorch.cpp]: -> getRouteNexthop

File: orchagent/vnetorch.cpp [hunk 8]
```diff
@@ -122,6 +124,112 @@ bool VNetVrfObject::updateObj(vector<sai_attribute_t>& attrs)
...
+    {
+        SWSS_LOG_INFO("VNET route '%s' does'nt exist", ipPrefix.to_string().c_str());
+        return false;
+    }
+
+    nh = routes_.at(ipPrefix);
+    return true;
+}
+
+sai_object_id_t VNetVrfObject::getNextHop(tunnelEndpoint& endp)
```
Comment (lguohan) [inline orchagent/vnetorch.cpp]: >getNextHop [](start = 31, length = 10)

-> getTunnelNexthop

File: orchagent/vnetorch.cpp [hunk 9]
```diff
@@ -122,6 +124,112 @@ bool VNetVrfObject::updateObj(vector<sai_attribute_t>& attrs)
...
+    nh_id = vxlan_orch->createNextHopTunnel(tun_name, endp.ip, endp.mac, endp.vni);
+    if (nh_id == SAI_NULL_OBJECT_ID)
+    {
+        throw std::runtime_error("NH Tunnel create failed for " + vnet_name_ + " ip " + endp.ip.to_string());
+    }
+
+    return nh_id;
+}
+
+bool VNetVrfObject::removeNextHop(tunnelEndpoint& endp)
```
Comment (lguohan) [inline orchagent/vnetorch.cpp]: removeTunnelNexthop

File: orchagent/vnetorch.cpp [hunk 10]
```diff
@@ -122,6 +124,112 @@ bool VNetVrfObject::updateObj(vector<sai_attribute_t>& attrs)
...
+{
+    auto tun_name = getTunnelName();
+
+    VxlanTunnelOrch* vxlan_orch = gDirectory.get<VxlanTunnelOrch*>();
+
+    if (!vxlan_orch->removeNextHopTunnel(tun_name, endp.ip, endp.mac, endp.vni))
+    {
+        SWSS_LOG_ERROR("VNET %s NH remove failed for '%s'", vnet_name_.c_str(), endp.ip.to_string().c_str());
```
Comment (lguohan) [inline orchagent/vnetorch.cpp]: >NH [](start = 32, length = 2)

-> tunnel nexthop

File: orchagent/vxlanorch.cpp [hunk 11]
```diff
@@ -390,20 +501,36 @@ VxlanTunnelOrch::createNextHopTunnel(string tunnelName, IpAddress& ipAddr, MacAd
...
+    if(!isTunnelExists(tunnelName))
+    {
+        SWSS_LOG_ERROR("Vxlan tunnel '%s' does not exists", tunnelName.c_str());
+        return true;
+    }
+
+    auto tunnel_obj = getVxlanTunnel(tunnelName);
+
+    //Delete request for the nh tunnel id
+    return tunnel_obj->removeNHTunnel(ipAddr, macAddress, vni);
```
Comment (lguohan) [inline orchagent/vxlanorch.cpp]: >removeNHTunnel [](start = 23, length = 14)

since it is already a tunnel object, you do not need tunnel in the function name. -> removeNexthop

File: orchagent/vxlanorch.h [hunk 12]
```diff
@@ -56,13 +104,23 @@ class VxlanTunnel
         return ids_.tunnel_encap_id;
     }
 
+    void updateNHTunnel(IpAddress& ipAddr, MacAddress macAddress, uint32_t vni, sai_object_id_t nh_id);
```
Comment (lguohan) [inline orchagent/vxlanorch.h]: -> updateNexthop

**Comment by lguohan** [comment]
>     sai_status_t status = sai_route_api->create_route_entry(&route_entry, 1, &route_attr);

we should probably increase the crm counter.

         if (route_entry.destination.addr_family == SAI_IP_ADDR_FAMILY_IPV4)
823         {
824             gCrmOrch->incCrmResUsedCounter(CrmResourceType::CRM_IPV4_ROUTE);
825         }
826         else
827         {
828             gCrmOrch->incCrmResUsedCounter(CrmResourceType::CRM_IPV6_ROUTE);
829         }

---
Refers to: orchagent/vnetorch.cpp:484 in 91d77df. [](commit_id = 91d77dfd07232bf326dad4167afe297d13b926ed, deletion_comment = False)

File: orchagent/intfsorch.cpp (hunk 1 above)
Comment (marian-pritsak) [reply orchagent/intfsorch.cpp]: But this vrf_id in this same function is to `addRouterIntfs(vrf_id, port)`
How does it work for rif and not for ip2me route?

Comment (prsunny) [reply orchagent/intfsorch.cpp]: `addRouterIntfs(vrf_id, port)` is a different flow (called from `VnetOrch`), `ip2Me ` route is not from `VnetOrch`. Remember, the IP address config doesn't have Vnetname in it.

File: orchagent/vnetorch.cpp (hunk 5 above)
Comment (prsunny) [reply orchagent/vnetorch.cpp]: The else case is handled by the Bitmap implementation. So it is left intentionally.
